docs/data/joy/guides/themeable-component/themeable-component.md

```diff
@@ -0,0 +1,309 @@
+
+```diff
++ import { useThemeProps } from '@mui/joy/styles';
+
+- const Stat = React.forwardRef(function Stat(props, ref) {
++ const Stat = React.forwardRef(function Stat(inProps, ref) {
++   const props = useThemeProps({ props: inProps, name: 'JoyStat' });
+    const { value, unit, ...other } = props;
+
+    return (
+      <StatRoot ref={ref} {...other}>
+        <StatValue>{value}</StatValue>
+        <StatUnit>{unit}</StatUnit>
+      </StatRoot>
+    );
+  });
+```
+
+Then your users can customize the default props of your component like this:
+
+```js
+import { extendTheme } from '@mui/joy/styles';
+
+const theme = extendTheme({
+  components: {
+    JoyStat: {
+      defaultProps: {
+        variant: 'outlined',
+      },
+    },
+  },
+});
+```
+
+## TypeScript
+
+If you use TypeScript, you should create interfaces for the component props and ownerState.
+
+```js
+interface StatProps {
+  value: number | string;
+  unit: string;
+  variant?: 'outlined';
+}
+
+interface StatOwnerState extends StatProps {
+  // …key value pairs for the internal state that you want to style the slot
+  // but don't want to expose to the users
+}
+```
+
+Then you can use them in the component and slots.
+
+```js
+const StatRoot = styled('div', {
+  name: 'JoyStat',
+  slot: 'root',
+})<{ ownerState: StatOwnerState }>(({ theme, ownerState }) => ({
+  display: 'flex',
+  flexDirection: 'column',
+  gap: theme.spacing(0.5),
+  padding: theme.spacing(3, 4),
+  backgroundColor: theme.palette.background.paper,
+  borderRadius: theme.shape.borderRadius,
+  boxShadow: theme.shadows[2],
+  letterSpacing: '-0.025em',
+  fontWeight: 600,
+  // typed-safe access to the `variant` prop
+  ...(ownerState.variant === 'outlined' && {
+    border: `2px solid ${theme.palette.divider}`,
+    boxShadow: 'none',
+  }),
+}));
+
+// …do the same for other slots
+
+const Stat = React.forwardRef<HTMLDivElement, StatProps>(function Stat(inProps, ref) {
+  const props = useThemeProps({ props: inProps, name: 'JoyStat' });
+  const { value, unit, variant, ...other } = props;
+
+  const ownerState = { ...props, variant };
+
+  return (
+    <StatRoot ref={ref} ownerState={ownerState} {...other}>
+      <StatValue ownerState={ownerState}>{value}</StatValue>
+      <StatUnit ownerState={ownerState}>{unit}</StatUnit>
+    </StatRoot>
+  );
+});
+```
+
+Finally, you have to add the Stat component the theme types.
+
+```ts
+import { Theme, StyleOverrides } from '@mui/joy/styles';
+import { StatProps, StatOwnerState } from '/path/to/Stat';
+
+declare module '@mui/joy/styles' {
+  interface Components {
+    JoyStat?: {
+      defaultProps?: Partial<StatProps>;
+      styleOverrides?: StyleOverrides<StatProps, StatOwnerState, Theme>;
+    };
+  }
+}
+```
+
+:::success
+🍾 Congratulations! You have successfully built a themeable component!.
+:::
+
+---
+
+## Template
+
+You can use the following template to create your own component. It is the full code of the `Stat` component in this guide.
```

Comment:
Reiterating the purpose of the template for anyone who skipped straight to it and might not be sure what they're looking at.

```suggestion
This template is the final product of the step-by-step guide above, demonstrating how to build a custom component that can be styled with the theme as if it was a built-in component.
```